llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp

```diff
@@ -0,0 +1,255 @@
+
+  auto transposedOp = builder.create<linalg::GenericOp>(
+      loc, outputTensor.getType(),
+      /*inputs=*/input, /*outputs=*/outputTensor, indexingMaps,
+      loopAttributeTypes,
+      [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+        nestedBuilder.create<linalg::YieldOp>(nestedLoc, args[0]);
+      });
+
+  return transposedOp.getResult(0);
+}
+
+// Rewrites iree_linalg_ext.pack to tensor.pad + tensor.expand_shape +
+// linalg.generic (transpose) ops.
+struct GeneralizePackOpPattern : OpRewritePattern<PackOp> {
+  using OpRewritePattern<PackOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(PackOp packOp,
+                                PatternRewriter &rewriter) const final {
+    if (!packOp.hasTensorSemantics()) {
+      return rewriter.notifyMatchFailure(packOp, "require tensor semantics");
+    }
+
+    Value input = getInputOrPaddedInput(rewriter, packOp);
+
+    SmallVector<ReassociationIndices> expandIndices;
+    SmallVector<int64_t> transPerm, innerPosAfterExpansion;
+    SmallVector<int64_t> inputVecCastShape;
+    ShapedType paddedInputType = input.getType();
+    int64_t inputRank = packOp.getInputRank();
+    SmallVector<bool> readInBounds(inputRank, true);
+    DenseMap<int64_t, OpFoldResult> tileAndPosMapping =
+        packOp.getDimAndTileMapping();
+    for (int64_t dim = 0, idx = 0; dim < inputRank; ++dim) {
+      int64_t dimSize = paddedInputType.getDimSize(dim);
+      if (!tileAndPosMapping.count(dim)) {
+        transPerm.push_back(inputVecCastShape.size());
+        innerPosAfterExpansion.push_back(-1);
+        inputVecCastShape.push_back(dimSize);
+        expandIndices.push_back(ReassociationIndices{idx});
+        idx++;
+        continue;
+      }
+
+      Optional<int64_t> tileSize =
+          getConstantIntValue(tileAndPosMapping.lookup(dim));
+      transPerm.push_back(inputVecCastShape.size());
+      inputVecCastShape.push_back(dimSize / tileSize.getValue());
+      innerPosAfterExpansion.push_back(inputVecCastShape.size());
+      inputVecCastShape.push_back(tileSize.getValue());
+      expandIndices.push_back(ReassociationIndices{idx, idx + 1});
+      idx += 2;
+    }
+
+    for (auto dim : extractFromI64ArrayAttr(packOp.getInnerDimsPos())) {
+      transPerm.push_back(innerPosAfterExpansion[dim]);
+    }
+    if (auto outerDims = packOp.getOuterDimsPerm()) {
+      transPerm = interchange(transPerm, extractFromI64ArrayAttr(outerDims));
+    }
+
+    Location loc = packOp.getLoc();
+    auto expandTargetType = RankedTensorType::get(
+        inputVecCastShape, paddedInputType.getElementType());
+    auto expand = rewriter.create<tensor::ExpandShapeOp>(loc, expandTargetType,
+                                                         input, expandIndices);
+    auto trans = createTransposeOp(loc, rewriter, expand, transPerm);
+    rewriter.replaceOp(packOp, trans);
+    return success();
+
+  }
+};
+
+struct ExpandShapeVectorizationPattern
```

Comment:
Can you please add some doc on this pattern?

Reply:
I added the comment, but it's not pushed... sorry about that. I'll push it to the branch now.